src/main/java/com/auth0/TransientCookieStore.java

```diff
@@ -0,0 +1,154 @@
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Allows storage and retrieval/removal of cookies.
+ */
+class TransientCookieStore {
+
+    private static final int MAX_AGE_SECONDS = 600; // 10 minutes
+
+    // Prevent instantiation
+    private TransientCookieStore() {}
+
+
+    /**
+     * Stores a state value as a cookie on the response.
+     *
+     * @param response the response object to set the cookie on
+     * @param state the value for the state cookie. If null, no cookie will be set.
+     * @param sameSite the value for the SameSite attribute on the cookie
+     * @param useLegacySameSiteCookie whether to set a fallback cookie or not
+     */
+    static void storeState(HttpServletResponse response, String state, SameSite sameSite, boolean useLegacySameSiteCookie) {
+        store(response, StorageUtils.STATE_KEY, state, sameSite, useLegacySameSiteCookie);
+    }
+
+    /**
+     * Stores a nonce value as a cookie on the response.
+     *
+     * @param response the response object to set the cookie on
+     * @param nonce the value for the nonce cookie. If null, no cookie will be set.
+     * @param sameSite the value for the SameSite attribute on the cookie
+     * @param useLegacySameSiteCookie whether to set a fallback cookie or not
+     */
+    static void storeNonce(HttpServletResponse response, String nonce, SameSite sameSite, boolean useLegacySameSiteCookie) {
+        store(response, StorageUtils.NONCE_KEY, nonce, sameSite, useLegacySameSiteCookie);
+    }
+
+    /**
+     * Gets the value associated with the state cookie and removes it.
+     * @param request the request object
+     * @param response the response object
+     * @param useLegacySameSiteCookie whether to use a fallback cookie or not
+     * @return the value of the state cookie, if it exists
+     */
+    static String getState(HttpServletRequest request, HttpServletResponse response, boolean useLegacySameSiteCookie) {
+        return getOnce(StorageUtils.STATE_KEY, request, response, useLegacySameSiteCookie);
+    }
+
+    /**
+     * Gets the value associated with the nonce cookie and removes it.
+     * @param request the request object
+     * @param response the response object
+     * @param useLegacySameSiteCookie whether to use a fallback cookie or not
+     * @return the value of the nonce cookie, if it exists
+     */
+    static String getNonce(HttpServletRequest request, HttpServletResponse response, boolean useLegacySameSiteCookie) {
+        return getOnce(StorageUtils.NONCE_KEY, request, response, useLegacySameSiteCookie);
+    }
+
+    private static void store(HttpServletResponse response, String key, String value, SameSite sameSite, boolean useLegacySameSiteCookie) {
+        Validate.notNull(response, "response must not be null");
+        Validate.notNull(key, "key must not be null");
+        Validate.notNull(sameSite, "sameSite must not be null");
+
+        if (value == null) {
```

Comment:
When would this happen? Is it ok to just return and do nothing or will there be unintended consequences? Feels to me like passing `null` to a storage function should tell me I'm doing something wrong.

Reply:
I understand what you mean here. Originally it did not accept null values as you suggest, but in another review feedback it was suggested to move the null checks out of `AuthorizeUrl` (before storing cookie values) and into here so that the logic in `AuthorizeUrl` was simplified. I'm fine with either way as the class is internal to the library.

Reply:
Up to you, just wanted to call it out. Sounds like it was thought through 👍